### PR TITLE
JDK-8297242: Use-after-free during library unloading on Linux

### PR DESCRIPTION
**Pull request by jcking**
ASan identified a use-after-free in `os::dll_unload` caused by referencing memory owned by the dynamic loader after closing the library.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297242](https://bugs.openjdk.org/browse/JDK-8297242): Use-after-free during library unloading on Linux

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11231/head:pull/11231` \
`$ git checkout pull/11231`

Update a local copy of the PR: \
`$ git checkout pull/11231` \
`$ git pull https://git.openjdk.org/jdk pull/11231/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11231`

View PR using the GUI difftool: \
`$ git pr show -t 11231`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11231.diff">https://git.openjdk.org/jdk/pull/11231.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/os/posix/os_posix.cpp [hunk 1]
```diff
@@ -716,9 +716,20 @@ void* os::dll_lookup(void* handle, const char* name) {
...
 void os::dll_unload(void *lib) {
-  const char* l_path = LINUX_ONLY(os::Linux::dll_path(lib))
-                       NOT_LINUX("<not available>");
-  if (l_path == NULL) l_path = "<not available>";
+  // os::Linux::dll_path returns a pointer to a string that is owned by the dynamic loader. Upon
+  // calling dlclose the dynamic loader may free the memory containing the string, thus we need to
+  // copy the string to be able to reference it after dlclose.
+  const char* l_path = NULL;
+  char* l_pathdup = NULL;
+  LINUX_ONLY(
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/os_posix.cpp:724]: The XXX_ONLY macros are intended for one-liners or in-line conditionals. For a block please use `#ifdef LINUX`.

Comment (jcking) [reply src/hotspot/os/posix/os_posix.cpp]: Ah, thanks for the correction. Was not sure. Fixed.

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks for fixing.

In case you are not aware hotspot changes require two reviews and a nominal 24 hour delay before integrating. Thanks.

**Pull request closed without merging**